File: archive/cdktf-ts/Pr6107/lib/database-stack.ts
import { Construct } from 'constructs';
import { RdsGlobalCluster } from '@cdktf/provider-aws/lib/rds-global-cluster';
import { RdsCluster } from '@cdktf/provider-aws/lib/rds-cluster';
import { RdsClusterInstance } from '@cdktf/provider-aws/lib/rds-cluster-instance';
import { DbSubnetGroup } from '@cdktf/provider-aws/lib/db-subnet-group';
import { SecurityGroup } from '@cdktf/provider-aws/lib/security-group';
import { SecurityGroupRule } from '@cdktf/provider-aws/lib/security-group-rule';
import { AwsProvider } from '@cdktf/provider-aws/lib/provider';
import { KmsKey } from '@cdktf/provider-aws/lib/kms-key';
import { KmsAlias } from '@cdktf/provider-aws/lib/kms-alias';

export interface DatabaseStackProps {
  environmentSuffix: string;
  primaryVpcId: string;
  drVpcId: string;
  primaryDbSubnetIds: string[];
  drDbSubnetIds: string[];
  primaryRegion: string;
  drRegion: string;
  primaryProvider?: AwsProvider;
  drProvider?: AwsProvider;
}

export interface DatabaseStackOutputs {
  globalClusterId: string;
  primaryClusterId: string;
  drClusterId: string;
  primaryClusterEndpoint: string;
  primaryClusterReaderEndpoint: string;
  drClusterEndpoint: string;
  drClusterReaderEndpoint: string;
}

export class DatabaseStack extends Construct {
  public readonly outputs: DatabaseStackOutputs;

  constructor(scope: Construct, id: string, props: DatabaseStackProps) {
    super(scope, id);

    const {
      environmentSuffix,
      primaryVpcId,
      drVpcId,
      primaryDbSubnetIds,
      drDbSubnetIds,
      primaryRegion,
      drRegion,
      primaryProvider,
      drProvider,
    } = props;

    const commonTags = {
      Environment: environmentSuffix,
      CostCenter: 'payment-processing',
      ManagedBy: 'cdktf',
    };

    // Security Groups
    const primaryDbSg = new SecurityGroup(this, 'primary-db-sg', {
      name: `payment-db-sg-${environmentSuffix}-${primaryRegion}`,
      description: 'Security group for Aurora database in primary region',
      vpcId: primaryVpcId,
      tags: {
        ...commonTags,
        'DR-Role': 'primary',
        Name: `payment-db-sg-${environmentSuffix}-${primaryRegion}`,
      },
      provider: primaryProvider,
    });

    new SecurityGroupRule(this, 'primary-db-ingress', {
      type: 'ingress',
      fromPort: 5432,
      toPort: 5432,
      protocol: 'tcp',
      cidrBlocks: ['10.0.0.0/16'],
      securityGroupId: primaryDbSg.id,
      description: 'PostgreSQL access from VPC',
      provider: primaryProvider,
    });

    new SecurityGroupRule(this, 'primary-db-egress', {
      type: 'egress',
      fromPort: 0,
      toPort: 0,
      protocol: '-1',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: primaryDbSg.id,
      description: 'Allow all outbound traffic',
      provider: primaryProvider,
    });

    const drDbSg = new SecurityGroup(this, 'dr-db-sg', {
      name: `payment-db-sg-${environmentSuffix}-${drRegion}`,
      description: 'Security group for Aurora database in DR region',
      vpcId: drVpcId,
      tags: {
        ...commonTags,
        'DR-Role': 'dr',
        Name: `payment-db-sg-${environmentSuffix}-${drRegion}`,
      },
      provider: drProvider,
    });

    new SecurityGroupRule(this, 'dr-db-ingress', {
      type: 'ingress',
      fromPort: 5432,
      toPort: 5432,
      protocol: 'tcp',
      cidrBlocks: ['10.1.0.0/16'],
      securityGroupId: drDbSg.id,
      description: 'PostgreSQL access from VPC',
      provider: drProvider,
    });

    new SecurityGroupRule(this, 'dr-db-egress', {
      type: 'egress',
      fromPort: 0,
      toPort: 0,
      protocol: '-1',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: drDbSg.id,
      description: 'Allow all outbound traffic',
      provider: drProvider,
    });

    // DB Subnet Groups
    const primarySubnetGroup = new DbSubnetGroup(this, 'primary-subnet-group', {
      name: `payment-db-subnet-${environmentSuffix}-${primaryRegion}`,
      subnetIds: primaryDbSubnetIds,
      tags: {
        ...commonTags,
        'DR-Role': 'primary',
      },
      provider: primaryProvider,
    });

    const drSubnetGroup = new DbSubnetGroup(this, 'dr-subnet-group', {
      name: `payment-db-subnet-${environmentSuffix}-${drRegion}`,
      subnetIds: drDbSubnetIds,
      tags: {
        ...commonTags,
        'DR-Role': 'dr',
      },
      provider: drProvider,
    });

    // Global Database Cluster
    const globalCluster = new RdsGlobalCluster(this, 'global-cluster', {
      globalClusterIdentifier: `payment-global-${environmentSuffix}`,
      engine: 'aurora-postgresql',
      engineVersion: '14.6',
      databaseName: 'payments',
      storageEncrypted: true,
      deletionProtection: false,
    });

    // Primary Regional Cluster
    const primaryCluster = new RdsCluster(this, 'primary-cluster', {
      clusterIdentifier: `payment-primary-${environmentSuffix}`,
      engine: 'aurora-postgresql',
      engineVersion: '14.6',
      databaseName: 'payments',
      masterUsername: 'dbadmin',
      masterPassword: `Payment${environmentSuffix}SecurePass123!`,
      dbSubnetGroupName: primarySubnetGroup.name,
      vpcSecurityGroupIds: [primaryDbSg.id],
      globalClusterIdentifier: globalCluster.id,
      backupRetentionPeriod: 7,
      preferredBackupWindow: '03:00-04:00',
      preferredMaintenanceWindow: 'sun:04:00-sun:05:00',
      skipFinalSnapshot: true,
      enabledCloudwatchLogsExports: ['postgresql'],
      tags: {
        ...commonTags,
        'DR-Role': 'primary',
      },
      provider: primaryProvider,
      dependsOn: [globalCluster],
    });

    // Primary cluster instances
    const primaryInstance1 = new RdsClusterInstance(
      this,
      'primary-instance-1',
      {
        identifier: `payment-primary-${environmentSuffix}-1`,
        clusterIdentifier: primaryCluster.id,
        instanceClass: 'db.r5.large',
        engine: 'aurora-postgresql',
        engineVersion: '14.6',
        publiclyAccessible: false,
        tags: {
          ...commonTags,
          'DR-Role': 'primary',
        },
        provider: primaryProvider,
      }
    );

    new RdsClusterInstance(this, 'primary-instance-2', {
      identifier: `payment-primary-${environmentSuffix}-2`,
      clusterIdentifier: primaryCluster.id,
      instanceClass: 'db.r5.large',
      engine: 'aurora-postgresql',
      engineVersion: '14.6',
      publiclyAccessible: false,
      tags: {
        ...commonTags,
        'DR-Role': 'primary',
      },
      provider: primaryProvider,
      dependsOn: [primaryInstance1],
    });

    // KMS Key for DR cluster encryption
    const drKmsKey = new KmsKey(this, 'dr-kms-key', {
      description: `KMS key for payment database DR cluster - ${environmentSuffix}`,
      enableKeyRotation: true,
      tags: {
        ...commonTags,
        'DR-Role': 'dr',
        Name: `payment-db-kms-${environmentSuffix}-${drRegion}`,
      },
      provider: drProvider,
    });

    new KmsAlias(this, 'dr-kms-alias', {
      name: `alias/payment-db-${environmentSuffix}-${drRegion}`,
      targetKeyId: drKmsKey.id,
      provider: drProvider,
    });

    // DR Regional Cluster (read-only replica)
    // Changed construct ID to force Terraform to treat this as a new resource
    // Also changed cluster identifier to avoid AWS name reservation conflicts
    const drCluster = new RdsCluster(this, 'dr-cluster-v2', {
      clusterIdentifier: `payment-dr-${environmentSuffix}-v2`,
      engine: 'aurora-postgresql',
      engineVersion: '14.6',
      dbSubnetGroupName: drSubnetGroup.name,
      vpcSecurityGroupIds: [drDbSg.id],
      globalClusterIdentifier: globalCluster.id,
      kmsKeyId: drKmsKey.arn,
      skipFinalSnapshot: true,
      enabledCloudwatchLogsExports: ['postgresql'],
      tags: {
        ...commonTags,
        'DR-Role': 'dr',
      },
      provider: drProvider,
      dependsOn: [primaryCluster],
      lifecycle: {
        createBeforeDestroy: false,
      },
    });

    // DR cluster instance
    new RdsClusterInstance(this, 'dr-instance-1-v2', {
      identifier: `payment-dr-${environmentSuffix}-v2-1`,
      clusterIdentifier: drCluster.id,
      instanceClass: 'db.r5.large',
      engine: 'aurora-postgresql',
      engineVersion: '14.6',
      publiclyAccessible: false,
      tags: {
        ...commonTags,
        'DR-Role': 'dr',
      },
      provider: drProvider,
    });

    this.outputs = {
      globalClusterId: globalCluster.id,
      primaryClusterId: primaryCluster.clusterIdentifier,
      drClusterId: drCluster.clusterIdentifier,
      primaryClusterEndpoint: primaryCluster.endpoint,
      primaryClusterReaderEndpoint: primaryCluster.readerEndpoint,
      drClusterEndpoint: drCluster.endpoint,
      drClusterReaderEndpoint: drCluster.readerEndpoint,
    };
  }
}

File: lib/stacks/database_stack.py
"""Database infrastructure - DynamoDB Global Tables and Aurora Global Database"""

import os
from constructs import Construct
from cdktf_cdktf_provider_aws.dynamodb_table import (
    DynamodbTable,
    DynamodbTableAttribute,
    DynamodbTableReplica,
    DynamodbTablePointInTimeRecovery,
    DynamodbTableGlobalSecondaryIndex,
)
from cdktf_cdktf_provider_aws.rds_cluster import RdsCluster
from cdktf_cdktf_provider_aws.rds_cluster_instance import RdsClusterInstance
from cdktf_cdktf_provider_aws.db_subnet_group import DbSubnetGroup
from cdktf_cdktf_provider_aws.rds_global_cluster import RdsGlobalCluster
from cdktf_cdktf_provider_aws.kms_key import KmsKey
from cdktf_cdktf_provider_aws.kms_alias import KmsAlias


class DatabaseStack(Construct):
    """Creates DynamoDB Global Tables and Aurora Global Database"""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        region: str,
        environment_suffix: str,
        is_primary: bool,
        network_stack=None
    ):
        super().__init__(scope, construct_id)

        self.region = region
        self.environment_suffix = environment_suffix
        self.is_primary = is_primary

        # KMS key for encryption
        self.kms_key = KmsKey(
            self,
            "db-kms-key",
            description=f"KMS key for database encryption in {region}",
            deletion_window_in_days=7,
            enable_key_rotation=True,
            tags={
                "Name": f"dr-db-kms-{region}-{environment_suffix}"
            }
        )

        KmsAlias(
            self,
            "db-kms-alias",
            name=f"alias/dr-db-{region}-{environment_suffix}",
            target_key_id=self.kms_key.id
        )

        # DynamoDB Global Table - only create in primary region with replica
        if self.is_primary:
            other_region = "us-east-2" if region == "us-east-1" else "us-east-1"
            # For Global Tables, omit kms_key_arn in replica to use AWS managed encryption
            replica_config = [
                DynamodbTableReplica(
                    region_name=other_region
                    # AWS managed encryption is used by default when kms_key_arn is omitted
                )
            ]

            self.dynamodb_table = DynamodbTable(
                self,
                "payments-table",
                name=f"dr-payments-{environment_suffix}",
                billing_mode="PAY_PER_REQUEST",
                hash_key="transactionId",
                range_key="timestamp",
                attribute=[
                    DynamodbTableAttribute(
                        name="transactionId",
                        type="S"
                    ),
                    DynamodbTableAttribute(
                        name="timestamp",
                        type="N"
                    ),
                    DynamodbTableAttribute(
                        name="customerId",
                        type="S"
                    )
                ],
                global_secondary_index=[
                    DynamodbTableGlobalSecondaryIndex(
                        name="CustomerIndex",
                        hash_key="customerId",
                        range_key="timestamp",
                        projection_type="ALL"
                    )
                ],
                replica=replica_config,
                point_in_time_recovery=DynamodbTablePointInTimeRecovery(
                    enabled=True
                ),
                stream_enabled=True,
                stream_view_type="NEW_AND_OLD_IMAGES",
                server_side_encryption={
                    "enabled": True
                    # Use AWS managed encryption for Global Tables
                    # Custom KMS requires valid ARN in each replica region
                },
                tags={
                    "Name": f"dr-payments-{environment_suffix}"
                }
            )
        else:
            # Secondary region doesn't create table - it's replicated from primary
            self.dynamodb_table = None

        # Aurora Global Database
        if self.is_primary:
            # Create global cluster in primary region only
            self.global_cluster = RdsGlobalCluster(
                self,
                "aurora-global-cluster",
                global_cluster_identifier=f"dr-aurora-global-{environment_suffix}",
                engine="aurora-postgresql",
                engine_version="14.6",
                database_name="payments",
                storage_encrypted=True,
                deletion_protection=False
            )

        # DB Subnet Group - get private subnets from network stack
        if network_stack is None:
            raise ValueError("network_stack is required for DatabaseStack")

        self.db_subnet_group = DbSubnetGroup(
            self,
            "aurora-subnet-group",
            name=f"dr-aurora-subnet-{region}-{environment_suffix}",
            subnet_ids=[subnet.id for subnet in network_stack.private_subnets],
            tags={
                "Name": f"dr-aurora-subnet-{region}-{environment_suffix}"
            }
        )

        # Aurora Cluster
        if self.is_primary:
            # Primary cluster (read-write)
            self.aurora_cluster = RdsCluster(
                self,
                "aurora-cluster",
                cluster_identifier=f"dr-aurora-{region}-{environment_suffix}",
                engine="aurora-postgresql",
                engine_version="14.6",
                engine_mode="provisioned",
                database_name="payments",
                master_username="dbadmin",
                master_password=os.environ.get("TF_VAR_db_password", "ChangeMe123!"),  # noqa: S105
                db_subnet_group_name=self.db_subnet_group.name,
                vpc_security_group_ids=[network_stack.aurora_security_group.id],
                backup_retention_period=7,
                preferred_backup_window="03:00-04:00",
                preferred_maintenance_window="mon:04:00-mon:05:00",
                storage_encrypted=True,
                kms_key_id=self.kms_key.arn,
                enabled_cloudwatch_logs_exports=["postgresql"],
                deletion_protection=False,
                skip_final_snapshot=True,
                global_cluster_identifier=self.global_cluster.id if self.is_primary else None,
                tags={
                    "Name": f"dr-aurora-{region}-{environment_suffix}"
                }
            )
        else:
            # Secondary cluster (read-only replica) joining global cluster
            # For cross-region replicas, do NOT specify master_username or master_password
            # The credentials are inherited from the primary cluster
            self.aurora_cluster = RdsCluster(
                self,
                "aurora-cluster",
                cluster_identifier=f"dr-aurora-{region}-{environment_suffix}",
                engine="aurora-postgresql",
                engine_version="14.6",
                engine_mode="provisioned",
                db_subnet_group_name=self.db_subnet_group.name,
                vpc_security_group_ids=[network_stack.aurora_security_group.id],
                storage_encrypted=True,
                kms_key_id=self.kms_key.arn,
                enabled_cloudwatch_logs_exports=["postgresql"],
                deletion_protection=False,
                skip_final_snapshot=True,
                global_cluster_identifier=f"dr-aurora-global-{environment_suffix}",
                tags={
                    "Name": f"dr-aurora-{region}-{environment_suffix}"
                }
            )

        # Aurora Cluster Instances (2 instances per cluster for HA)
        for i in range(2):
            RdsClusterInstance(
                self,
                f"aurora-instance-{i}",
                identifier=f"dr-aurora-{region}-{i}-{environment_suffix}",
                cluster_identifier=self.aurora_cluster.id,
                instance_class="db.r6g.large",
                engine="aurora-postgresql",
                engine_version="14.6",
                publicly_accessible=False,
                tags={
                    "Name": f"dr-aurora-{region}-{i}-{environment_suffix}"
                }
            )
